File: benchmark/go/StressTest.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"time"
)

type SimpleTimer struct {
	t1 int64
	t2 int64
}

func (t *SimpleTimer) reset() int64 {
	t.t1 = t.t2
	t.t2 = time.Now().UnixNano() / 1000000

	elapsed := t.t2 - t.t1
	return elapsed
}

func newTimer() SimpleTimer {
	t := SimpleTimer{}
	t.reset()
	t.t1 = t.t2
	return t
}

const (
	_1K = int(1024)
	_1M = _1K * _1K
)

var (
	g_testNodes []*BaggageNode
	g_testTitle string
)

type StressTestOptions struct {
	basePayload        int
	primitiveBytes     int
	circularRefPercent int
	replacePercent     int
}

type StressTest struct {
	_cntAllocatedNodes int
	_options           *StressTestOptions
}

type Node struct {
	_id        int32
	var1       *Node
	var2       *Node
	var3       *Node
	parent     *Node
	root       *Node
	primitives []int8
}

type BaggageNode struct {
	Node
	_cntNodes       int
	_cntCircularRef int
	_test           *StressTest
}

func newNode(test *StressTest, parent *Node) *Node {
	node := &Node{}
	node.initNode(test, parent)
	return node
}

func (self *Node) initNode(test *StressTest, parent *Node) {
	self._id = int32(test._cntAllocatedNodes)
	test._cntAllocatedNodes++
	self.primitives = make([]int8, test._options.primitiveBytes)
	if parent != nil {
		self.parent = parent
		self.root = parent.root
	}
}

func (self *BaggageNode) initBaggageNode(test *StressTest, cntNodes int) {
	self.initNode(test, nil)
	self._test = test
	self._cntNodes = cntNodes
	self._cntCircularRef = test._options.circularRefPercent * cntNodes / 100
	self.root = &self.Node
	self.attachChildren(&self.Node, 0)
}

func (self *BaggageNode) attachChildren(node *Node, layer int) {
	node.var1 = self.createChildNode(node, layer)
	node.var2 = self.createChildNode(node, layer)
	node.var3 = self.createChildNode(node, layer)
}

func (self *BaggageNode) createChildNode(parent *Node, layer int) *Node {
	idx_child := self._test._cntAllocatedNodes - int(self._id)
	if idx_child >= self._cntNodes {
		return nil
	}

	if idx_child >= self._cntCircularRef {
		parent = nil
	}
	node := newNode(self._test, parent)
	layer++
	if layer < 7 {
		self.attachChildren(node, layer)
	}
	return node
}

func (self *StressTest) init(options *StressTestOptions) {
	self._cntAllocatedNodes = 0
	self._options = options
}

func (self *StressTest) createBaggage(cntNodes int) *BaggageNode {
	if cntNodes == 0 {
		return nil
	}
	root := &BaggageNode{}
	root.initBaggageNode(self, cntNodes)
	return root
}

func (self *StressTest) stressTest(payload int) int64 {
	defer onOOM()

	fmt.Println("--------------------------------------")
	fmt.Println("Preparing test...")

	baggages := make([]*BaggageNode, payload)
	g_testNodes = baggages
	self._cntAllocatedNodes = 0
	runtime.GC()
	runtime.GC()

	for i := int(0); i < payload; i++ {
		baggages[i] = self.createBaggage(_1K)
	}

	timer := newTimer()
	timer2 := newTimer()

	fmt.Println(g_testTitle + " StressTest started. " +
		"nodes: " + strconv.FormatInt(int64(self._cntAllocatedNodes)/1024, 10) + "k")

	timer.reset()
	timer2.reset()
	updateLimit := payload * self._options.replacePercent / 100
	if updateLimit == 0 {
		updateLimit = 1
	}
	for repeat := 0; repeat < 5; repeat++ {
		for i := int(0); i < payload; i++ {
			idx := i % updateLimit
			baggages[idx] = self.createBaggage(_1K)
		}

		fmt.Println(strconv.Itoa(repeat+1) + ") " + strconv.FormatInt(timer.reset(), 10) + " ms")
	}

	baggages = nil
	elapsed := timer2.reset()
	fmt.Println("Total elapsed time: " + strconv.FormatInt(elapsed, 10))
	return elapsed
}

func main() {
	options := &StressTestOptions{}
	options.basePayload, _ = strconv.Atoi(os.Args[1])
	options.primitiveBytes, _ = strconv.Atoi(os.Args[2])
	options.circularRefPercent, _ = strconv.Atoi(os.Args[3])
	options.replacePercent, _ = strconv.Atoi(os.Args[4])

	g_testTitle = "[" + runtime.Version() + "]"

	fmt.Println(g_testTitle + " StressTest")
	fmt.Println(" - Base payload: " + strconv.Itoa(options.basePayload) + "k nodes")
	fmt.Println(" - Primitive bytes: " + strconv.Itoa(options.primitiveBytes) + " byte")
	fmt.Println(" - Circular ref percent: " + strconv.Itoa(options.circularRefPercent) + "%")
	fmt.Println(" - Replace percent: " + strconv.Itoa(options.replacePercent) + "%")

	test := &StressTest{}
	test.init(options)

	payload := 0
	print(g_testTitle)
	for {
		payload += int(options.basePayload)
		elapsed := test.stressTest(payload)
		print(", " + strconv.Itoa(int(elapsed)))
	}
}

func onOOM() {
	if err := recover(); err != nil {
		fmt.Println("termnated by OutOfMemoryError")
	}
}

File: benchmark/dotnet/src/StressTest.cs
using System;

#pragma warning disable CS0169
#pragma warning disable CS0414
#pragma warning disable CS0649

namespace gctest
{
    public class StressTest {

        const int _1K = 1024;
        const int _1M = _1K * _1K;
        public int _cntAllocatedNodes;
        public Options _options;
        public static String g_testTitle;

        public struct Options {
            public int basePayload;
            public int primitiveBytes;
            public int circularRefPercent;
            public int replacePercent;
        };

        class Node { 
            public int _id;
            public Node var1;
            public Node var2;
            public Node var3;
            public Node parent;
            public Node root;
    		public Object primitives;
            
            public Node(StressTest test, Node parent) {
                this._id = test._cntAllocatedNodes ++;
                this.primitives = new byte[test._options.primitiveBytes];
                if (parent != null) {
                    this.parent = parent;
                    this.root = parent.root;
                }
            }
        };

        class BaggageNode : Node {
            int _cntNodes; 
            int _cntCircularRef;
            StressTest _test;

            public BaggageNode(StressTest test, int cntNodes) : base(test, null) {
                this._test = test;
                this._cntNodes = cntNodes;
                this._cntCircularRef = test._options.circularRefPercent * cntNodes / 100;
                this.root = this;
                attachChildren(this, 0);
            }

            void attachChildren(Node node, int layer) {
                node.var1 = createChildNode(node, layer);
                node.var2 = createChildNode(node, layer);
                node.var3 = createChildNode(node, layer);
            }

            Node createChildNode(Node parent, int layer) {
                long idx_child = _test._cntAllocatedNodes - this._id;
                if (idx_child >= _cntNodes) {
                    return null;
                }

                if (idx_child >= _cntCircularRef) {
                    parent = null;
                }
                Node node = new Node(_test, parent);
                if (++layer < 7) {
                    attachChildren(node, layer);
                }
                return node;
            }
        };        

        StressTest(Options options) {
            this._cntAllocatedNodes = 0;
            this._options = options;
        }

        Node createBaggage(int cntNodes) {
            if (cntNodes <= 0) {
                return null;
            }
            BaggageNode root = new BaggageNode(this, cntNodes);
            return root;
        }

        static Object[] g_testNodes;
        public long stressTest(int payload) {
            Console.WriteLine("--------------------------------------\n");
            Console.WriteLine("Preparing test...");
            SimpleTimer timer = new SimpleTimer();
            SimpleTimer timer2 = new SimpleTimer();

            Object[] baggages = new Object[payload];
            g_testNodes = baggages;
            _cntAllocatedNodes = 0;
            GC.Collect();
            GC.Collect();

            for (int i = 0; i < payload; i++) {
                baggages[i] = createBaggage(_1K);
            }
            Console.WriteLine(g_testTitle + " StressTest started. " + 
                "nodes: " + _cntAllocatedNodes/1024 + "k");

            timer.reset();
            timer2.reset();
            int updateLimit = payload * _options.replacePercent / 100;
            if (updateLimit == 0) {
                updateLimit = 1;
            }
            for (int repeat = 0; repeat++ < 5; ) {
                for (int i = 0; i < payload; i++) {
                    int idx = i % updateLimit;
                    baggages[idx] = createBaggage(_1K);
                }
                Console.WriteLine(repeat + ") " + timer.reset() + " ms");
            }

            long elapsed = timer2.reset();
            Console.WriteLine("Total elapsed time: " + elapsed);
            return elapsed;
        }

        public static void doTest(Options options) {
            
            String gcMode = System.Runtime.GCSettings.IsServerGC ? "Server" : "Workstation";
            g_testTitle = "[.Net v" + Environment.Version.ToString() + "/" + gcMode + "]";

            Console.WriteLine(g_testTitle + " StressTest");
            Console.WriteLine(" - Base payload: " + options.basePayload + "k nodes");
		    Console.WriteLine(" - Primitive bytes: " + options.primitiveBytes + " byte");
            Console.WriteLine(" - Circular ref percent: " + options.circularRefPercent + "%");
            Console.WriteLine(" - Replace percent: " + options.replacePercent + "%");
            Console.WriteLine(" - TotalAvailableMemory: " + GC.GetGCMemoryInfo().TotalAvailableMemoryBytes/_1M + "m");

            try {
                Console.Error.Write("#" + g_testTitle);
                StressTest test = new StressTest(options);
                for (int payload = 0;;) {
                    payload += options.basePayload;
                    long elapsed = test.stressTest(payload);
                    Console.Error.Write(", " + elapsed);
                    Console.Error.Flush();
                }
            }
            catch (OutOfMemoryException) {
                Console.WriteLine("\nTest is terminated by OutOfMemoryException");
            }
        }
    }
}